src/icp_v2.cc

```diff
@@ -536,15 +502,40 @@ doV2Query(int fd, Ip::Address &from, char *buf, icp_common_t header)
 #endif /* USE_ICMP */
 
     /* The peer is allowed to use this cache */
-    ICP2State *state = new ICP2State(header, icp_request);
-    state->fd = fd;
-    state->from = from;
-    state->url = xstrdup(url);
-    state->flags = flags;
-    state->rtt = rtt;
-    state->src_rtt = src_rtt;
+    ICP2State state(header, icp_request);
+    state.fd = fd;
+    state.from = from;
+    state.url = xstrdup(url);
+    state.flags = flags;
+    state.rtt = rtt;
+    state.src_rtt = src_rtt;
+
+    const auto e = storeGetPublic(url, Http::METHOD_GET);
+
+    icp_opcode codeToSend;
+
+    if (e && state.confirmAndPrepHit(*e)) {
```

Comment:
Judging by the current use cases, confirmAndPrepHit(e) should stay private and we should add a public boolean isHit(void) method that will call storeGetPublic(), confirmAndPrepHit(), and `e->abandon()`, all without "leaking" `e` outside.